### PR TITLE
AP_Landing: Touchdown yaw correction with altitude or weight on wheels

### PR DESCRIPTION
**Pull request by EShamaev**
Superseded https://github.com/ArduPilot/ardupilot/pull/7898

Side wind 15 m/s

Without:
image

With:
image

Adresses: #7867

### CODE REVIEW

File: libraries/AP_Landing/AP_Landing_Slope.cpp [hunk 1]
```diff
@@ -66,6 +68,22 @@ bool AP_Landing::type_slope_verify_land(const Location &prev_WP_loc, Location &n
...
+    // Check for weight-on-wheels touchdown
+    if (!touched_down) {
+        const bool touchdown_alt_override = !is_zero(touchdown_altitude) &&
+                (height < touchdown_altitude);
+        const bool touchdown_debounced = (AP_LandingGear::instance()->get_wow_state() == AP_LandingGear::LG_WOW) &&
+                AP_LandingGear::instance()->get_wow_state_duration_ms() >= 500;
+
+        if (touchdown_alt_override || touchdown_debounced) {
+            touched_down = true; // sticky flag
+            type_slope_stage = SLOPE_STAGE_FINAL; // force flare
```
Comment (magicrub) [inline libraries/AP_Landing/AP_Landing_Slope.cpp]: I don't like that you trigger the flare like this. It messes up the announcement stuff later on. This state should only be set in one place. So, let's make a flag to force the flare. Here's a commit I just wrote that you can cherry pick
https://github.com/magicrub/ardupilot/commit/c888b6ca7e309c39d0e6937cfa7ea5deb0c7c4dc

File: libraries/AP_Landing/AP_Landing.h [hunk 2]
```diff
@@ -134,6 +134,10 @@ class AP_Landing {
     adjusted_relative_altitude_cm_fn_t adjusted_relative_altitude_cm_fn;
     disarm_if_autoland_complete_fn_t disarm_if_autoland_complete_fn;
     update_flight_stage_fn_t update_flight_stage_fn;
+    
+    // saved bearing for yaw correction after touchdown
+    float runway_bearing;
+    bool touched_down;
```
Comment (magicrub) [inline libraries/AP_Landing/AP_Landing.h]: can you move touched_down up into the flags struct?

File: libraries/AP_Landing/AP_Landing_Slope.cpp (hunk 1 above)
Comment (EShamaev) [reply libraries/AP_Landing/AP_Landing_Slope.cpp]: Done

File: libraries/AP_Landing/AP_Landing.h (hunk 2 above)
Comment (EShamaev) [reply libraries/AP_Landing/AP_Landing.h]: Done

**Comment by tridge** [comment]
@EShamaev I've reverted this PR as it caused a really bizarre landing in some cases, and in one case in SITL I saw a fly-away
See the issue here: https://github.com/ArduPilot/ardupilot/issues/9993
once we work out the cause and fix it we can put this feature back in

**Comment by EShamaev** [comment]
@tridge Bug was found, runway bearing was not calculated early enough. Please advise how to fix it. Should I make a completely new PR with the fix or Make only the PR for this change before revertion?

**Comment by magicrub** [comment]
Can you add the reverted changes to PR 10080?

**Pull request merged**